bindings/src/main/java/io/pravega/storage/extendeds3/ExtendedS3Storage.java

```diff
@@ -194,24 +194,36 @@ public boolean supportsTruncation() {
      * Iterator for segments in ExtendedS3Storage.
      */
     public class ExtendedS3SegmentIterator implements Iterator<SegmentProperties> {
-        ListObjectsResult results;
-        S3Object current;
-        Iterator<S3Object> innerIterator;
-        java.util.function.Predicate<S3Object> patternMatchPredicate;
+        private final java.util.function.Predicate<S3Object> patternMatchPredicate;
+        private ListObjectsResult results;
+        private Iterator<S3Object> innerIterator;
+        private S3Object current;
+        private boolean isAvailable;
 
         ExtendedS3SegmentIterator(java.util.function.Predicate<S3Object> patternMatchPredicate) {
             this.results = client.listObjects(config.getBucket(), config.getPrefix());
             this.innerIterator = results.getObjects().iterator();
             this.patternMatchPredicate = patternMatchPredicate;
+            this.isAvailable = false;
         }
 
+        /**
+         * Method to check the presence of next element in the iterator.
+         * It also sets the position of the current element for Next method, but repetitive call to this method before Next
+         * will not advance the current element.
+         * @return true if the next element is there, else false.
+         */
         @Override
         public boolean hasNext() {
+            if (isAvailable) { // If the method was already called before next method call, return true if last call returned true.
```

Comment:
I've been thinking. ...

This could be achieved by doign something like 
(constructor)
```
innerIterator = client.listObjects(config.getBucket(), config.getPrefix()).stream()
                      .flatMap(r -> r.getObjects().stream())
                      .filter(patternMatchIPredicate)
                      .map(this::toSegmentProperties)
                      .iterator()
```

`hasNext`: `return innerIterator.hasNext()`
`next`: `return innerIterator.next()`
`toSegmentProperties`: put the code you have in `next` here

Does this work? It may require a couple of adjustments but it should eliminate the need for all this extra code.
Please fix the other iterators too.

Reply:
Ideall the `toSegmentProperties` would share code with `doGetStreamSegmentInfo`. Both places extract info from S3ObjectMetadata into our own types.